File: demo/src/main/java/com/example/demo/model/Donation.java
package com.example.demo.model;

import java.time.LocalDate;

import com.fasterxml.jackson.annotation.JsonBackReference;

import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "donation")
public class Donation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private LocalDate date;
    private String hospital;
    private int units;
    private String notes;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "donor_id")
    @JsonBackReference
    private Donor donor;
    
}

File: front/lib/models/donation.dart
class Donation {
  final int? id;
  final int donorId;
  final String? date;
  final String? hospital;
  final int? units;
  final String? notes;
  final int donationCount;


  Donation({
    this.id,
    required this.donorId,
    required this.date,
    required this.hospital,
    required this.units,
    required this.notes,
    this.donationCount = 0,
  });

  factory Donation.fromJson(Map<String, dynamic> json) {
  return Donation(
    id: json['id'] as int?,
    date: json['date']?.toString(),
    hospital: json['hospital']?.toString(),
    units: json['units'] as int?,
    notes: json['notes']?.toString(),
    donorId: json['donor'] != null ? (json['donor'] is Map ? json['donor']['id'] : json['donor']) as int : 0, 
  );
}

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'date': date,
      'hospital': hospital,
      'units': units,
      'notes': notes,
      'donor': {'id': donorId}
    };
  }
}



